drivers/uacpi/BindsIo.cpp

```diff
@@ -0,0 +1,226 @@
+    {
+        (void)address; (void)offset; (void)byte_width; (void)value;
+
+        Log("Attempted raw PCI read", LogLevel::Warning);
+        return UACPI_STATUS_UNIMPLEMENTED;
+    }
+
+    uacpi_status uacpi_kernel_pci_write(uacpi_pci_address* address, uacpi_size offset, uacpi_u8 byte_width, uacpi_u64 value)
+    {
+        (void)address; (void)offset; (void)byte_width; (void)value;
+
+        Log("Attempted raw PCI read", LogLevel::Warning);
+        return UACPI_STATUS_UNIMPLEMENTED;
+    }
+
+    union IoMapping //:meme:
+    {
+        struct
+        {
+            uint16_t base;
+            uint16_t length;
+        };
+        uintptr_t packed;
+    };
+
+    uacpi_status uacpi_kernel_io_map(uacpi_io_addr base, uacpi_size len, uacpi_handle* out_handle)
+    {
+#ifdef __x86_64__
+        VALIDATE_(base <= 0xFFFF, UACPI_STATUS_INVALID_ARGUMENT);
+        VALIDATE_(len <= 0xFFFF, UACPI_STATUS_INVALID_ARGUMENT);
+
+        IoMapping mapping { .base = (uint16_t)base, .length = (uint16_t)len };
+        *out_handle = reinterpret_cast<uacpi_handle>(mapping.packed);
+
+        return UACPI_STATUS_OK;
+#else
+        return UACPI_STATUS_UNIMPLEMENTED;
+#endif
+    }
+
+    void uacpi_kernel_io_unmap(uacpi_handle handle)
+    { (void)handle; } //no-op
+
+    uacpi_status uacpi_kernel_io_read(uacpi_handle handle, uacpi_size offset, uacpi_u8 byte_width, uacpi_u64* value)
+    {
+#ifdef __x86_64__
+        VALIDATE_(handle != nullptr, UACPI_STATUS_INVALID_ARGUMENT);
+        VALIDATE_(offset <= 0xFFFF, UACPI_STATUS_INVALID_ARGUMENT);
+        VALIDATE_(byte_width <= 4 && byte_width != 3, UACPI_STATUS_INVALID_ARGUMENT);
+        VALIDATE_(value != nullptr, UACPI_STATUS_INVALID_ARGUMENT);
+
+        const IoMapping mapping { .packed = reinterpret_cast<uintptr_t>(handle) };
+        const uint16_t port = mapping.base + offset;
+
+        *value = PortRead(port, byte_width);
+        return UACPI_STATUS_OK;
+#else
+        return UACPI_STATUS_UNIMPLEMENTED;
+#endif
+    }
+
+    uacpi_status uacpi_kernel_io_write(uacpi_handle handle, uacpi_size offset, uacpi_u8 byte_width, uacpi_u64 value)
+    {
+#ifdef __x86_64__
+        VALIDATE_(handle != nullptr, UACPI_STATUS_INVALID_ARGUMENT);
+        VALIDATE_(offset <= 0xFFFF, UACPI_STATUS_INVALID_ARGUMENT);
+        VALIDATE_(byte_width <= 4 && byte_width != 3, UACPI_STATUS_INVALID_ARGUMENT);
+
+        const IoMapping mapping { .packed = reinterpret_cast<uintptr_t>(handle) };
+        const uint16_t port = mapping.base + offset;
+
+        PortWrite(port, byte_width, value);
+        return UACPI_STATUS_OK;
+#else
+        return UACPI_STATUS_UNIMPLEMENTED;
+#endif
+    }
+
+    void* uacpi_kernel_map(uacpi_phys_addr addr, uacpi_size len)
+    {
+        if (len == 0)
+            return nullptr;
+
+        const npk_vm_flags flags = (npk_vm_flags)(VmMmio | VmWrite);
+        return npk_vm_alloc(len, (void*)addr, flags, nullptr);
+    }
+
+    void uacpi_kernel_unmap(void *addr, uacpi_size len)
```

Comment:
isn't len unused? You just check the value, but is not used in the free process.

Reply:
Yeah it's unused. This call ends up at the vmm, which tracks the length internally, so it's not needed for `npk_vm_unmap`. Maybe wiuldnt hurt to add a comment for this :)